fallback.go

```diff
@@ -43,22 +43,27 @@ func New(trace plugin.Handler) (f *Fallback) {
 func (f Fallback) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
 	nw := nonwriter.New(w)
 	rcode, err := plugin.NextOrFailure(f.Name(), f.Next, ctx, nw, r)
-	if err != nil {
-		return rcode, err
-	}
-	// PROXY ALWAYS RETURN 0
-	if rcode == 0 {
-		rcode = nw.Msg.Rcode
+
+	//By default the rules_index is equal rcode, so in such way we handle the case
+	//when rcode is SERVFAIL and nw.Msg is nil, otherwise we use nw.Msg.Rcode
+	//because, for example, for the following cases like NXDOMAIN, REFUSED the rcode is 0 (returned by proxy)
+	//A proxy doesn't return 0 only in case SERVFAIL
+	rules_index := rcode
+	if nw.Msg != nil {
+		rules_index = nw.Msg.Rcode
 	}
-	if u, ok := f.rules[rcode]; ok {
+
+	if u, ok := f.rules[rules_index]; ok {
 		p := f.proxy.New(f.trace, u)
 		if p == nil {
-			return rcode, fmt.Errorf("cannot create fallback proxy")
+			return rules_index, fmt.Errorf("cannot create fallback proxy")
```

Comment:
I guess if we throw an error, we should return SERVFAIL .. no ? Also because initial writer w is not filled.

Reply:
I guess - yes, and this is what we do, in case of SERVFAIL the rcode is SERVFAIL as well as rules_index, isn't it?

Reply:
hum.. no. you can have a rules_index that is REFUSE, you catch if in your fallback rule, call the f.proxy and get a NIL.

=> in that case, I would expect you to return SERVFAIL, "cannot create fallback", instead of REFUSE, "cannot create fallback".

Reply:
Oh, got it, you meant we need to return 

return dns.RcodeServerFailure, fmt.Errorf("cannot create fallback proxy")?

Reply:
yes.